CONTRIBUTING.md

```diff
@@ -0,0 +1,91 @@
+# Contributing to itertools
+
+We use stable Rust only.
+Please check the minimum version of Rust we use in `Cargo.toml`.
+
+To pass CI tests, you should run `cargo fmt` and it might become mandatory that
+your code changes are free of "compile warnings" and "clippy warnings"
+(see [#720](https://github.com/rust-itertools/itertools/pull/720) and [#740](https://github.com/rust-itertools/itertools/pull/740)).
+
+No guideline about using `#[inline]` yet.
+
+## `core::iter::Iterator` vs `itertools::Itertools`
+For new features, please first consider filing a PR to [rust-lang/rust](https://github.com/rust-lang/rust),
+adding your new feature to the `Iterator` trait of the standard library, if you believe it is reasonable.
+If it isn't accepted there, proposing it for inclusion in `itertools` is a good idea.
+The reason for doing is this is so that we avoid future breakage as with `.flatten()`.
+However, if your feature involves heap allocation, such as storing elements in a `Vec<T>`,
+then it can't be accepted into `libcore`, and you should propose it for `itertools` directly instead.
+
+## Write a new method/adaptor for `Itertools` trait
+### Behind `use_std`/`use_alloc` feature?
+If it needs the "std" then it should be behind the `use_std` feature,
+or if it requires heap allocation (such as using vectors) then it should be behind the `use_alloc` feature.
+Otherwise it should be able to run in `no_std` context.
+
+This mostly applies to your new module, each import from it, and to your new `Itertools` method.
+
+### Pick the right receiver
+`self`, `&mut self` or `&self`? From [#710](https://github.com/rust-itertools/itertools/pull/710):
+
+- Take by value when:
+    - It transfers ownership to another iterator type, such as `filter`, `map`...
+    - It consumes the iterator completely, such as `count`, `last`, `max`...
+- Mutably borrow when it consumes only part of the iterator, such as `find`, `all`, `try_collect`...
+- Immutably borrow when there is no change, such as `size_hint`.
+
+## Specialize `Iterator` methods
+It might be more performant to specialize some methods.
+However, each specialization should be thoroughly tested.
```

Comment:
I think we should add something like: *Correctly specializing methods can be difficult, and we do not require that you do it on your initial PR.*

What do you think about this policy, @phimuemue?

Reply:
@jswrenn I like your idea. (It implicitly aids review by keeping PRs small.)

Should the guidelines mention that we have a canonical way to test specialized methods (in order to avoid a hundred hand-rolled implementations)?

Reply:
jswrenn's addition being approved is now committed.
Nothing yet about tests/benchmarks.